### PR TITLE
Add low-memory interface for building, sending and appending messages

### PR DESCRIPTION
**Pull request by disaykin**

### CODE REVIEW

File: src/core/rfc822/MCMessageBuilder.cpp [hunk 1]
```diff
@@ -810,6 +810,37 @@ Data * MessageBuilder::dataForEncryption()
...
+{
+    int col;
+    bool success = true;
+
+    col = 0;
+    struct mailmime * mime = mimeAndFilterBccAndForEncryption(false, false);
+
+    FILE * f = fopen(filename->fileSystemRepresentation(), "w");
+    if (f) {
+        mailmime_write_driver(doWriteToFile, f, &col, mime);
```
Comment (dinhviethoa) [inline src/core/rfc822/MCMessageBuilder.cpp]: You could probably use mailmime_write_file() here.

File: src/core/rfc822/MCMessageBuilder.cpp [hunk 2]
```diff
@@ -810,6 +810,37 @@ Data * MessageBuilder::dataForEncryption()
...
+bool MessageBuilder::storeToFile(String * filename)
+{
+    int col;
+    bool success = true;
+
+    col = 0;
+    struct mailmime * mime = mimeAndFilterBccAndForEncryption(false, false);
+
+    FILE * f = fopen(filename->fileSystemRepresentation(), "w");
+    if (f) {
```
Comment (dinhviethoa) [inline src/core/rfc822/MCMessageBuilder.cpp]:
```
FILE * f = fopen();
if (f == NULL) {
  goto free_mime;
}

int col = 0;
r = mailmime_write_file(f, &col, mime);
if (r != NULL) {
  goto close_f;
}
fclose(f);
mailmime_free(mime);

return true;

close_f:
fclose(f);
free_mime:
mailmime_free(mime);
return false;
```

File: src/core/rfc822/MCMessageBuilder.h [hunk 3]
```diff
@@ -40,7 +40,10 @@ namespace mailcore {
         
         virtual Data * data();
         virtual Data * dataForEncryption();
-        
+
+        // Store builded message to file. Return false in case of failure.
+        virtual bool storeToFile(String * filename);
```
Comment (dinhviethoa) [inline src/core/rfc822/MCMessageBuilder.h]: Could you rename all storeToFile() to writeToFile() and return a ErrorCode instead?
(to be consistent with Data::writeToFile()).

Thanks!

File: src/objc/rfc822/MCOMessageBuilder.mm [hunk 4]
```diff
@@ -70,6 +70,11 @@ - (NSData *) dataForEncryption
     return MCO_OBJC_BRIDGE_GET(dataForEncryption);
 }
 
+- (BOOL) storeToFile:(NSString *)filename
```
Comment (dinhviethoa) [inline src/objc/rfc822/MCOMessageBuilder.mm]: - (BOOL) writeToFile:(NSString *)filename error:(NSError **)errorPtr

File: src/objc/smtp/MCOSMTPSession.mm [hunk 5]
```diff
@@ -170,6 +170,25 @@ - (MCOSMTPSendOperation *) sendOperationWithData:(NSData *)messageData
     return result;
 }
 
+- (MCOSMTPSendOperation *) sendOperationWithDataFile:(NSString *)messageDataFile
```
Comment (dinhviethoa) [inline src/objc/smtp/MCOSMTPSession.mm]: I'd like to call this method:
- (MCOSMTPSendOperation *) sendOperationWithContentsOfFile:(NSString *)path
  from:(MCOAddress *)from
  recipients:(NSArray *)recipients

**Comment by dinhviethoa** [comment]
I'd like to get rid of the parameter `sendAsIs`.
Here's my thoughts:

we could add a Data * overridenHeaderData or a MessageHeader * overridenHeader parameter to the following method:

```
void SMTPSession::sendMessage(Address * from, Array * recipients, Data * messageData,
    SMTPProgressCallback * callback, ErrorCode * pError)
```

```
void SMTPSession::sendMessage(Address * from, Array * recipients, Data * messageData,
    MessageHeader * overridenHeader,
    SMTPProgressCallback * callback, ErrorCode * pError)
```

add in libetpan, the proper calls to be able to send headers and body separately.
Additional methods similar to `mailesmtp_send()` and `mailesmtp_send_quit()` that would allow that.
Let me know if I can help with the libetpan part.

As for the following:
```
- (MCOSMTPSendOperation *) sendOperationWithContentsOfFile:(NSString *)path from:(MCOAddress *)from recipients:(NSArray *)recipients
```

could you create the equivalent method in MCSMTPAsyncSession?
I'd like to have a 1:1 mapping of Objective-C-C++ methods.

File: src/objc/imap/MCOIMAPSession.mm [hunk 6]
```diff
@@ -264,6 +264,23 @@ - (MCOIMAPAppendMessageOperation *)appendMessageOperationWithFolder:(NSString *)
     return MCO_TO_OBJC_OP(coreOp);
 }
 
+- (MCOIMAPAppendMessageOperation *)appendMessageOperationWithFolder:(NSString *)folder
```
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPSession.mm]: Could you create the C++ method equivalent instead?

File: src/objc/imap/MCOIMAPSession.h [hunk 7]
```diff
@@ -305,6 +305,21 @@
...
+
+     MCOIMAPOperation * op = [session appendMessageOperationWithFolder:@"Sent Mail" messageDataFile:rfc822DataFilename flags:MCOMessageFlagNone customFlags:@[@"$CNS-Greeting-On"]];
+     [op start:^(NSError * __nullable error, uint32_t createdUID) {
+       if (error == nil) {
+         NSLog(@"created message with UID %lu", (unsigned long) createdUID);
+       }
+     }];
+ */
+- (MCOIMAPAppendMessageOperation *)appendMessageOperationWithFolder:(NSString *)folder
+                                                    messageDataFile:(NSString *)messageDataFile
```
Comment (dinhviethoa) [inline src/objc/imap/MCOIMAPSession.h]: contentsAtPath:(NSString *)path

**Comment by disaykin** [comment]
Done

**Comment by dinhviethoa** [comment]
That's brilliant!
We could even make the Bcc case consume less memory by tweaking libetpan.

**Pull request merged**